README.md

```diff
@@ -0,0 +1,111 @@
+# stream-json-rpc
+Call remote process methods easily, using the transport that you need.
+
+## Usage
+Example using node-ipc
+```typescript
+// The transport layer must implement the `stream.Duplex` interface
+// Here are the ones for `node-ipc`
+import { Duplex } from 'stream';
+
+class ServerDuplex extends Duplex {
+  ipc: NodeIPC.Server; // instance of nodeipc.server
+  socket: NodeIPC.Socket; // destination socket
+
+  constructor(ipc: NodeIPC.Server, socket: NodeIPC.Socket) {
+    super();
+    this.ipc = ipc;
+    this.socket = socket;
+
+    ipc.on('data', data => {
+      this.push(data);
+    });
+  }
+
+  // tslint:disable-next-line
+  _write(chunk: any, _encoding: any, callback: any) {
+    this.ipc.emit(this.socket, chunk);
+    callback();
+  }
+
+  // tslint:disable-next-line
+  _read(_size: any) {}
+}
+
+class ClientDuplex extends Duplex {
+  ipc: NodeIPC.Client; // ipc socket opened between current process and the server
+
+  constructor(ipc: NodeIPC.Client) {
+    super();
+    this.ipc = ipc;
+
+    ipc.on('data', data => {
+      this.push(data);
+    });
+  }
+
+  // tslint:disable-next-line
+  _write(chunk: any, _encoding: any, callback: any) {
+    this.ipc.emit(chunk);
+    callback();
+  }
+
+  // tslint:disable-next-line
+  _read(_size: any) {}
+}
+```
+
+And the actual implementation would look like:
+##### Process 1
+```typescript
+import { rpcchannel } from 'stream-json-rpc';
+
+// This process acts as node-ipc server
+// But it actually being a "server" is not really relevant,
+// as any process can call any other process if they can directly communicate.
+
+const ipc = getIPC(); // Instance of NodeIPC.Server
+const sockets = new Map(); // List of clients
+const channel = rpcchannel(); // Create the channel on the server side
+
+// At first connection, store the client socket
+const firstConnection = (data: Buffer, socket: any) => {
+  const id = data.toString('utf-8');
+  sockets.set(id, socket);
+  channel.setLink(id, new ServerDuplex(ipc, socket));
+  ipc.off('data', firstConnection);
+};
+ipc.on('data', firstConnection);
+
+// Register handlers
+channel.addRequestHandler('inc', ({ value }: any) => {
+  return value + 1;
+});
+```
+
+##### Process 2
+```typescript
+import { rpcchannel } from 'stream-json-rpc';
+
+// This process acts as node-ipc client
+
+// Instance of NodeIPC.Client, connected to the server
+// Also, getIPC here should send a first message with its id to the server.
+// That way the server can finish initializing the connection with this process.
+// (i.e. call `channel.setLink` on his side)
+const ipcClient = getIPC();
+const channel = rpcchannel();
+channel.setLink('server', new TestDuplex(ipcClient));
+
+// Call remote method on process 1
+channel
+  .request('server', 'inc', {
```

Comment:
I'm not sure to understand yet what is `server`. I assume it's the destination. I don't think it's necessary for the user to know where the request should go.

I think this should go above.

Something like:
```js
const channel = rpcchannel();
const peer = channel.getPeer('server', new TestDuplex(ipcClient));

peer.request('inc', {...})
```

Reply:
Even more, I think we can decouple the implementation of wire from RPCPeer implementation:
```
import Peer from 'json-rpc-peer';
const channel = rpcchannel();
const duplex = channel.stream('server', new TestDuplex(ipcClient));

const peer = new Peer();
duplex.pipe(peer)
peer.pipe(duplex)

peer.request('inc', {...})
```

Reply:
I'm not sure what you mean by "need to handle message itself" and why the second comment is not ok.

Here's an example where `Peer` handles message with current version of `json-rpc-peer`.
```js
import Peer from 'json-rpc-peer';
const channel = rpcchannel();
const duplex = channel.getStream('server', new TestDuplex(ipcClient));

const peer = new Peer((mesg) => {
  if (msg.method == 'ping') {
    return true;
  }
});
duplex.pipe(peer)
peer.pipe(duplex)

peer.request('inc', {...})
```

Though I think `json-rpc-peer` should have a better way to attach handlers. See my 2nd comment above.